### DISCUSSION TITLE
Any good ideas how to render marks correctly?

### DISCUSSION THREAD

**Discussion by gitcatrat** [ueberdosis/tiptap#3377] [General Questions] [answered]
Hi 👋 

I need to render editor JSON output without importing editor packages .

I've worked with several rich text editor document models before but this structure caught me by surprise:

Identical marks are spread around to multiple text nodes, makes it kind of tricky to lift them up and "merge" text under common marks. In my experience common attributes are usually already lifted to common ancestor in document output. This is often not an issue with marks like bold and italic (can be duplicated in DOM), but 3 side-by-side links looks super weird with hover states, etc.

Naive solution produces this output (**see JSON below**):
```html
<p>
  <a href="http://localhost/"><em>Go </em></a>
  <a href="http://localhost/">to </a>
  <a href="http://localhost/"><strong>homepage</strong></a>
</p>
```

How it should look like:
```html
<p>
  <a href="http://localhost/">
     <em>Go </em>
     to 
     <strong>homepage</strong>
  </a>
</p>
```

Here's a basic version 0.1 I tried but it's not generic enough: https://jsfiddle.net/wmucoz49/
Any suggestions to "lift common marks up" and wrap texts using them? Any generic utility functions that I couldn't find?

---

Example document (only relevant part for readability):
```json
{
      "type": "paragraph",
      "content": [
        {
          "type": "text",
          "marks": [
            {
              "type": "link",
              "attrs": {
                "href": "http://localhost/",
                "target": "_blank",
                "class": null
              }
            },
            {
              "type": "italic"
            }
          ],
          "text": "Go "
        },
        {
          "type": "text",
          "marks": [
            {
              "type": "link",
              "attrs": {
                "href": "http://localhost/",
                "target": "_blank",
                "class": null
              }
            }
          ],
          "text": "to "
        },
        {
          "type": "text",
          "marks": [
            {
              "type": "link",
              "attrs": {
                "href": "http://localhost/",
                "target": "_blank",
                "class": null
              }
            },
            {
              "type": "bold"
            }
          ],
          "text": "homepage"
        }
      ]
    }
```

**Comment by gitcatrat** [comment, marked as answer]
Took some inspiration from ProseMirror renderer and adjusted it to my rendering framework.
https://github.com/ProseMirror/prosemirror-model/blob/master/src/to_dom.ts#L46

**Reply by gethari** [reply]
Would u please share the solution? If you could , would help others to wandering here

**Reply by kasvith** [reply]
@gitcatrat any update on your solution?

**Reply by kasvith** [reply]
Prosemirror seems to have worst schema ever :( i also got into same problem

**Reply by kasvith** [reply]
@gitcatrat you can also try remirror...which has a good react renderer also
